File: sdk/verifiers/rust/src/lifecycle.rs
// Copyright 2026 Pipelock contributors
// SPDX-License-Identifier: Apache-2.0

use crate::types::{ChainResult, Receipt};
use std::collections::HashMap;

const MAX_SAFE_INTEGER: u64 = 9_007_199_254_740_991;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct LifecycleReport {
    pub status: String,
    pub reason: String,
}

#[derive(Default)]
struct RunState {
    opened: bool,
    closed: bool,
    intents: HashMap<String, u64>,
    outcomes: HashMap<String, u64>,
    saw_heartbeat: bool,
    last_heartbeat: u64,
    heartbeat_gap: bool,
    last_durability: Option<DurabilitySnapshot>,
    durability_drop: bool,
}

#[derive(Clone, Copy)]
struct DurabilitySnapshot {
    fsync_errors_gated: Option<u64>,
    durability_blocks: Option<u64>,
}

// analyze_lifecycle mirrors the Go verifier's packet-level lifecycle verdict.
// It deliberately reports an open but otherwise valid run as LIMITED rather
// than rejecting the packet: integrity and a clean session close answer
// different questions.
pub fn analyze_lifecycle(receipts: &[Receipt], chain: &ChainResult) -> LifecycleReport {
    if receipts.is_empty() {
        return report("UNVERIFIED", "no_receipts");
    }
    let missing_open = !chain.valid
        && chain.integrity_verified
        && chain.failure_kind.as_deref() == Some("lifecycle_missing_open");
    if !chain.valid && !missing_open {
        return report("BROKEN", "chain_broken");
    }
    if missing_open {
        return report("UNVERIFIED", "no_open");
    }
    if !receipts
        .iter()
        .any(|receipt| session_control(receipt).is_some())
    {
        return report("UNVERIFIED", "no_lifecycle");
    }

    let mut runs = HashMap::<String, RunState>::new();
    for receipt in receipts {
        let Some(record) = receipt.get("action_record") else {
            continue;
        };
        let control = session_control(receipt);
        let run_nonce = run_nonce_for(record, control);
        if run_nonce.is_empty() {
            return report("BROKEN", "chain_broken");
        }
        let state = runs.entry(run_nonce).or_default();
        match record
            .get("decision_phase")
            .and_then(serde_json::Value::as_str)
        {
            Some("intent") => increment(
                &mut state.intents,
                record
                    .get("action_id")
                    .and_then(serde_json::Value::as_str)
                    .unwrap_or(""),
            ),
            Some("outcome") => increment(
                &mut state.outcomes,
                record
                    .get("action_id")
                    .and_then(serde_json::Value::as_str)
                    .unwrap_or(""),
            ),
            _ => {}
        }

        let Some(control) = control else {
            continue;
        };
        match control.get("kind").and_then(serde_json::Value::as_str) {
            Some("session_open") => {
                state.opened = true;
                state.closed = false;
            }
            Some("session_close") => {
                let Some(close) = control.get("close").and_then(serde_json::Value::as_object)
                else {
                    return report("BROKEN", "chain_broken");
                };
                let action_run_nonce = record
                    .get("run_nonce")
                    .and_then(serde_json::Value::as_str)
                    .unwrap_or("");
                if action_run_nonce.is_empty()
                    || close.get("run_nonce").and_then(serde_json::Value::as_str)
                        != Some(action_run_nonce)
                {
                    return report("BROKEN", "chain_broken");
                }
                state.closed = true;
                if !apply_durability(state, control.get("close")) {
                    return report("BROKEN", "chain_broken");
                }
            }
            Some("heartbeat") | Some("session_heartbeat") => {
                let Some(heartbeat) = control
                    .get("heartbeat")
                    .and_then(serde_json::Value::as_object)
                else {
                    return report("BROKEN", "chain_broken");
                };
                let action_run_nonce = record
                    .get("run_nonce")
                    .and_then(serde_json::Value::as_str)
                    .unwrap_or("");
                if action_run_nonce.is_empty()
                    || heartbeat
                        .get("run_nonce")
                        .and_then(serde_json::Value::as_str)
                        != Some(action_run_nonce)
                {
                    return report("BROKEN", "chain_broken");
                }
                let beat = heartbeat.get("beat").map(lifecycle_counter);
                let Some(beat) = beat else {
                    if !apply_durability(state, control.get("heartbeat")) {
                        return report("BROKEN", "chain_broken");
                    }
                    continue;
                };
                let Some(beat) = beat else {
                    return report("BROKEN", "chain_broken");
                };
                if (state.saw_heartbeat && beat != state.last_heartbeat + 1)
                    || (!state.saw_heartbeat && beat != 1)
                {
                    state.heartbeat_gap = true;
                }
                state.saw_heartbeat = true;
                state.last_heartbeat = beat;
                if !apply_durability(state, control.get("heartbeat")) {
                    return report("BROKEN", "chain_broken");
                }
            }
            _ => {}
        }
    }

    runs.values()
        .map(assess_run)
        .reduce(worse)
        .unwrap_or_else(|| report("UNVERIFIED", "no_lifecycle"))
}

fn apply_durability(state: &mut RunState, value: Option<&serde_json::Value>) -> bool {
    let snapshot = DurabilitySnapshot {
        fsync_errors_gated: match value.and_then(|value| value.get("fsync_errors_gated")) {
            Some(value) => match lifecycle_counter(value) {
                Some(value) => Some(value),
                None => return false,
            },
            None => None,
        },
        durability_blocks: match value.and_then(|value| value.get("durability_blocks")) {
            Some(value) => match lifecycle_counter(value) {
                Some(value) => Some(value),
                None => return false,
            },
            None => None,
        },
    };
    if let Some(previous) = state.last_durability {
        if snapshot
            .fsync_errors_gated
            .zip(previous.fsync_errors_gated)
            .is_some_and(|(current, previous)| current < previous)
            || snapshot
                .durability_blocks
                .zip(previous.durability_blocks)
                .is_some_and(|(current, previous)| current < previous)
        {
            state.durability_drop = true;
        }
    }
    state.last_durability = Some(snapshot);
    true
}

fn lifecycle_counter(value: &serde_json::Value) -> Option<u64> {
    value.as_u64().filter(|value| *value <= MAX_SAFE_INTEGER)
}

fn session_control(receipt: &Receipt) -> Option<&serde_json::Map<String, serde_json::Value>> {
    receipt
        .get("action_record")
        .and_then(|record| record.get("session_control"))
        .and_then(serde_json::Value::as_object)
}

fn run_nonce_for(
    record: &serde_json::Value,
    control: Option<&serde_json::Map<String, serde_json::Value>>,
) -> String {
    if let Some(run_nonce) = record.get("run_nonce").and_then(serde_json::Value::as_str) {
        if !run_nonce.is_empty() {
            return run_nonce.to_string();
        }
    }
    let Some(control) = control else {
        return String::new();
    };
    let payload = match control.get("kind").and_then(serde_json::Value::as_str) {
        Some("session_open") => control.get("open"),
        Some("heartbeat") | Some("session_heartbeat") => control.get("heartbeat"),
        Some("session_close") => control.get("close"),
        _ => None,
    };
    payload
        .and_then(|payload| payload.get("run_nonce"))
        .and_then(serde_json::Value::as_str)
        .unwrap_or("")
        .to_string()
}

fn increment(values: &mut HashMap<String, u64>, key: &str) {
    *values.entry(key.to_string()).or_default() += 1;
}

fn assess_run(state: &RunState) -> LifecycleReport {
    if !state.opened {
        // An observed run nonce without its own session_open contradicts the
        // transcript. UNVERIFIED/no_open is reserved for the chain verifier's
        // explicit, integrity-proven lifecycle_missing_open failure kind above.
        return report("BROKEN", "chain_broken");
    }
    if state.outcomes.iter().any(|(action_id, outcomes)| {
        *outcomes > 0 && state.intents.get(action_id).unwrap_or(&0) == &0
    }) {
        return report("BROKEN", "chain_broken");
    }
    if state
        .intents
        .iter()
        .any(|(action_id, intents)| state.outcomes.get(action_id).unwrap_or(&0) < intents)
    {
        return report("LIMITED", "open_action");
    }
    if state.durability_drop {
        return report("BROKEN", "chain_broken");
    }
    if state.heartbeat_gap {
        return report("LIMITED", "heartbeat_gap");
    }
    if !state.closed {
        return report("LIMITED", "abnormal_end");
    }
    report("LIMITED", "bounded_closed")
}

fn report(status: &str, reason: &str) -> LifecycleReport {
    LifecycleReport {
        status: status.to_string(),
        reason: reason.to_string(),
    }
}

fn worse(current: LifecycleReport, next: LifecycleReport) -> LifecycleReport {
    let current_key = (
        status_severity(&current.status),
        reason_severity(&current.reason),
    );
    let next_key = (status_severity(&next.status), reason_severity(&next.reason));
    if next_key > current_key {
        next
    } else {
        current
    }
}

fn status_severity(status: &str) -> u8 {
    match status {
        "BROKEN" => 3,
        "UNVERIFIED" => 2,
        "LIMITED" => 1,
        _ => 0,
    }
}

fn reason_severity(reason: &str) -> u8 {
    match reason {
        "chain_broken" => 100,
        "no_open" => 80,
        "no_lifecycle" | "recorder_disabled" | "no_receipts" => 70,
        "open_action" => 50,
        "heartbeat_gap" => 40,
        "abnormal_end" => 30,
        "bounded_closed" => 10,
        _ => 0,
    }
}

File: sdk/verifiers/ts/src/lifecycle.ts
// Copyright 2026 Pipelock contributors
// SPDX-License-Identifier: Apache-2.0

import type { ChainResult, LifecycleReason, LifecycleStatus, Receipt } from "./types.js";

export interface LifecycleReport {
  status: LifecycleStatus;
  reason: LifecycleReason;
}

interface RunState {
  opened: boolean;
  closed: boolean;
  intents: Map<string, number>;
  outcomes: Map<string, number>;
  sawHeartbeat: boolean;
  lastHeartbeat: number;
  heartbeatGap: boolean;
  lastDurability?: DurabilitySnapshot;
  durabilityDrop: boolean;
}

interface DurabilitySnapshot {
  fsyncErrorsGated?: number;
  durabilityBlocks?: number;
}

const limited: LifecycleStatus = "LIMITED";
const broken: LifecycleStatus = "BROKEN";
const unverified: LifecycleStatus = "UNVERIFIED";

// analyzeLifecycle mirrors the Go verifier's packet-level lifecycle verdict.
// It deliberately reports an open but otherwise valid run as LIMITED rather
// than rejecting the packet: integrity and a clean session close answer
// different questions.
export function analyzeLifecycle(receipts: Receipt[], chain: ChainResult): LifecycleReport {
  if (receipts.length === 0) return { status: unverified, reason: "no_receipts" };
  const missingOpen =
    !chain.valid &&
    chain.integrity_verified === true &&
    chain.failure_kind === "lifecycle_missing_open";
  if (!chain.valid && !missingOpen) return { status: broken, reason: "chain_broken" };
  if (missingOpen) return { status: unverified, reason: "no_open" };
  const controls = receipts.filter((receipt) => sessionControl(receipt) !== undefined);
  if (controls.length === 0) return { status: unverified, reason: "no_lifecycle" };

  const runs = new Map<string, RunState>();
  const stateFor = (runNonce: string): RunState => {
    let state = runs.get(runNonce);
    if (state === undefined) {
      state = {
        opened: false,
        closed: false,
        intents: new Map<string, number>(),
        outcomes: new Map<string, number>(),
        sawHeartbeat: false,
        lastHeartbeat: 0,
        heartbeatGap: false,
        durabilityDrop: false,
      };
      runs.set(runNonce, state);
    }
    return state;
  };

  for (const receipt of receipts) {
    const record = receipt.action_record;
    if (record === undefined) continue;
    const control = sessionControl(receipt);
    const runNonce = runNonceFor(record, control);
    if (runNonce === "") return { status: broken, reason: "chain_broken" };
    const state = stateFor(runNonce);

    if (record.decision_phase === "intent") increment(state.intents, record.action_id ?? "");
    if (record.decision_phase === "outcome") increment(state.outcomes, record.action_id ?? "");

    if (control === undefined) continue;
    const kind = control["kind"];
    if (kind === "session_open") {
      state.opened = true;
      state.closed = false;
      continue;
    }
    if (kind === "session_close") {
      const close = objectField(control, "close");
      if (
        close === undefined ||
        typeof record.run_nonce !== "string" ||
        record.run_nonce === "" ||
        close["run_nonce"] !== record.run_nonce
      ) {
        return { status: broken, reason: "chain_broken" };
      }
      state.closed = true;
      if (!applyDurability(state, close)) {
        return { status: broken, reason: "chain_broken" };
      }
      continue;
    }
    if (kind === "heartbeat" || kind === "session_heartbeat") {
      const heartbeat = objectField(control, "heartbeat");
      if (
        heartbeat === undefined ||
        typeof record.run_nonce !== "string" ||
        record.run_nonce === "" ||
        heartbeat["run_nonce"] !== record.run_nonce
      ) {
        return { status: broken, reason: "chain_broken" };
      }
      const beat = heartbeat?.["beat"];
      if (beat !== undefined && !validCounter(beat)) {
        return { status: broken, reason: "chain_broken" };
      }
      if (beat !== undefined) {
        if (
          (state.sawHeartbeat && beat !== state.lastHeartbeat + 1) ||
          (!state.sawHeartbeat && beat !== 1)
        ) {
          state.heartbeatGap = true;
        }
        state.sawHeartbeat = true;
        state.lastHeartbeat = beat;
      }
      if (!applyDurability(state, heartbeat)) return { status: broken, reason: "chain_broken" };
    }
  }

  let report: LifecycleReport | undefined;
  for (const state of runs.values()) {
    report = worse(report, assessRun(state));
  }
  return report ?? { status: unverified, reason: "no_lifecycle" };
}

function applyDurability(state: RunState, value: Record<string, unknown> | undefined): boolean {
  const fsyncErrorsGated = counter(value?.["fsync_errors_gated"]);
  const durabilityBlocks = counter(value?.["durability_blocks"]);
  if (fsyncErrorsGated === null || durabilityBlocks === null) return false;
  const snapshot: DurabilitySnapshot = {
    fsyncErrorsGated,
    durabilityBlocks,
  };
  if (
    state.lastDurability !== undefined &&
    ((snapshot.fsyncErrorsGated !== undefined &&
      state.lastDurability.fsyncErrorsGated !== undefined &&
      snapshot.fsyncErrorsGated < state.lastDurability.fsyncErrorsGated) ||
      (snapshot.durabilityBlocks !== undefined &&
        state.lastDurability.durabilityBlocks !== undefined &&
        snapshot.durabilityBlocks < state.lastDurability.durabilityBlocks))
  ) {
    state.durabilityDrop = true;
  }
  state.lastDurability = snapshot;
  return true;
}

function validCounter(value: unknown): value is number {
  return typeof value === "number" && Number.isSafeInteger(value) && value >= 0;
}

function counter(value: unknown): number | undefined | null {
  if (value === undefined) return undefined;
  return validCounter(value) ? value : null;
}

function sessionControl(receipt: Receipt): Record<string, unknown> | undefined {
  const value = receipt.action_record?.session_control;
  return value !== null && value !== undefined && typeof value === "object" && !Array.isArray(value)
    ? value
    : undefined;
}

function objectField(
  value: Record<string, unknown>,
  key: string,
): Record<string, unknown> | undefined {
  const field = value[key];
  return field !== null && typeof field === "object" && !Array.isArray(field)
    ? (field as Record<string, unknown>)
    : undefined;
}

function runNonceFor(
  record: NonNullable<Receipt["action_record"]>,
  control: Record<string, unknown> | undefined,
): string {
  if (typeof record.run_nonce === "string" && record.run_nonce !== "") return record.run_nonce;
  if (control === undefined) return "";
  const kind = control["kind"];
  const payload =
    kind === "session_open"
      ? objectField(control, "open")
      : kind === "heartbeat" || kind === "session_heartbeat"
        ? objectField(control, "heartbeat")
        : kind === "session_close"
          ? objectField(control, "close")
          : undefined;
  return typeof payload?.["run_nonce"] === "string" ? (payload["run_nonce"] as string) : "";
}

function increment(values: Map<string, number>, key: string): void {
  values.set(key, (values.get(key) ?? 0) + 1);
}

function assessRun(state: RunState): LifecycleReport {
  // An observed run nonce without its own session_open contradicts the
  // transcript. UNVERIFIED/no_open is reserved for the chain verifier's
  // explicit, integrity-proven lifecycle_missing_open failure kind above.
  if (!state.opened) return { status: broken, reason: "chain_broken" };
  for (const [actionID, outcomes] of state.outcomes) {
    if ((state.intents.get(actionID) ?? 0) === 0 && outcomes > 0)
      return { status: broken, reason: "chain_broken" };
  }
  for (const [actionID, intents] of state.intents) {
    if ((state.outcomes.get(actionID) ?? 0) < intents)
      return { status: limited, reason: "open_action" };
  }
  if (state.durabilityDrop) return { status: broken, reason: "chain_broken" };
  if (state.heartbeatGap) return { status: limited, reason: "heartbeat_gap" };
  if (!state.closed) return { status: limited, reason: "abnormal_end" };
  return { status: limited, reason: "bounded_closed" };
}

function worse(current: LifecycleReport | undefined, next: LifecycleReport): LifecycleReport {
  if (current === undefined) return next;
  const statusOrder: Record<LifecycleStatus, number> = {
    LIMITED: 1,
    UNVERIFIED: 2,
    BROKEN: 3,
  };
  const reasonOrder: Record<LifecycleReason, number> = {
    chain_broken: 100,
    no_open: 80,
    no_lifecycle: 70,
    recorder_disabled: 70,
    no_receipts: 70,
    open_action: 50,
    heartbeat_gap: 40,
    abnormal_end: 30,
    bounded_closed: 10,
  };
  if (statusOrder[next.status] > statusOrder[current.status]) return next;
  if (
    statusOrder[next.status] === statusOrder[current.status] &&
    reasonOrder[next.reason] > reasonOrder[current.reason]
  ) {
    return next;
  }
  return current;
}
